src/org/exist/storage/XQueryPool.java

```diff
@@ -146,23 +147,19 @@ private void returnObject(final Source source, final CompiledXQuery xquery) {
             if (pool.size() < maxPoolSize) {
                 source.setCacheTimestamp(System.currentTimeMillis());
 
-                final Deque<CompiledXQuery> stack = pool.computeIfAbsent(source, (k) -> new ArrayDeque<>());
+                final Set<CompiledXQuery> stack = pool.computeIfAbsent(source, (k) -> new ConcurrentSkipListSet<>());
 
                 if (stack.size() < maxStackSize) {
-                    // check if the query is already in pool before adding,
-                    // may happen for modules, don't add it a second time!
-                    if(!stack.contains(xquery)) {
-                        stack.push(xquery);
-                    }
+                    stack.add(xquery);
```

Comment:
@shabanovd I fear there is a race between lines 173 and 176:

1) Thread A determines the query is invalid and is then suspended
2) Thread B causes the query to become valid and is then suspended
3) Thread A continues, and removes the valid compiled queries from the stack,